### PR TITLE
Add new unit test for configurable-route control loop.

### PR DESCRIPTION
**Pull request by gcs278**
Add new unit test for configurable-route control loop. 

Achieves about 70% code coverage. Tests almost all functions including the Reconcile function. Adds common functions to fake a controller-runtime client, controller-runtime cache, and a client-go clientset.

- `pkg/operator/controller/configurable-route/controller_test.go`: New Unit tests for configurable-route control loop.
- `test/unit/fake_client.go`: Added common fake client initialization functions.

### CODE REVIEW

**Comment by gcs278** [comment]
@candita I know this is low priority, but have you gotten a chance to look at it? Any questions?

**Comment by candita** [comment]
> @candita I know this is low priority, but have you gotten a chance to look at it? Any questions?

Sorry for the delay, will try today.

**Comment by candita** [comment]
Did you miss adding go.mod and go.sum?

File: pkg/operator/controller/configurable-route/controller_test.go [hunk 1]
```diff
@@ -0,0 +1,899 @@
...
+			if tc.ingressConfig != nil {
+				if err := r.client.Create(context.Background(), tc.ingressConfig); err != nil {
+					t.Errorf("error creating ingress config: %v", err)
+				}
+			}
+
+			if _, err := r.Reconcile(context.Background(), tc.request); err == nil && tc.expectError {
+				t.Errorf("expected error, got no error")
+			} else if err != nil && !tc.expectError {
+				t.Errorf("did not expected error: %v", err)
```
Comment (candita) [inline pkg/operator/controller/configurable-route/controller_test.go:159]:
```suggestion
				t.Errorf("did not expect error: %v", err)
```

File: pkg/operator/controller/configurable-route/controller_test.go [hunk 2]
```diff
@@ -0,0 +1,899 @@
...
+				} else {
+					t.Errorf("error retrieving roles from client: %v", err)
+				}
+			}
+		})
+	}
+}
+
+// Test_intersectingComponentRoutes verifies intersectingComponentRoutes
+func Test_intersectingComponentRoutes(t *testing.T) {
```
Comment (candita) [inline pkg/operator/controller/configurable-route/controller_test.go:214]: This function is already a part of Reconcile.  Does it give us additional value to unit test it outside of a unit test for Reconcile?

File: pkg/operator/controller/configurable-route/controller_test.go [hunk 3]
```diff
@@ -0,0 +1,899 @@
...
+							// Value that we want to ignore in the comparison.
+							actualRole.ResourceVersion = ""
+							actualRole.GenerateName = ""
+							expectedRole.GenerateName = ""
+							assert.Equal(t, expectedRole, actualRole)
+						}
+					}
+				} else {
+					t.Errorf("error retrieving roles from client: %v", err)
+				}
```
Comment (candita) [inline pkg/operator/controller/configurable-route/controller_test.go:184]: Can you encapsulate these sections around the `List` calls to the api and make them helper functions?

File: pkg/operator/controller/configurable-route/controller_test.go [hunk 4]
```diff
@@ -0,0 +1,899 @@
...
+							}
+						}
+						if assert.NotEmptyf(t, actualRoleBinding, fmt.Sprintf("Could not find expected rolebinding: %s/%s", expectedRoleBinding.Namespace, expectedRoleBinding.Name)) {
+							// Value that we want to ignore in the comparison.
+							expectedRoleBinding.ResourceVersion = ""
+							assert.Equal(t, expectedRoleBinding, actualRoleBinding)
+						}
+					}
+				} else {
+					t.Errorf("error retrieving roles from client: %v", err)
```
Comment (candita) [inline pkg/operator/controller/configurable-route/controller_test.go:206]: Same for this section.

File: pkg/operator/controller/configurable-route/controller_test.go [hunk 5]
```diff
@@ -0,0 +1,899 @@
...
+			},
+			roles: rbacv1.RoleList{
+				Items: []rbacv1.Role{
+					*newRoleForServiceCert("foo-role", "openshift-config", "bar-hash", ""),
+				},
+			},
+			expectedRoles: rbacv1.RoleList{},
+		},
+		{
+			name:           "No existing hashes list",
```
Comment (candita) [inline pkg/operator/controller/configurable-route/controller_test.go:380]: `go fmt` on this file?

File: pkg/operator/controller/configurable-route/controller_test.go [hunk 6]
```diff
@@ -0,0 +1,899 @@
...
+						}
+					}
+					if assert.NotEmptyf(t, actualRole, fmt.Sprintf("Could not find expected role: %s/%s", expectedRole.Namespace, expectedRole.Name)) {
+						// Value that we want to ignore in the comparison.
+						actualRole.ResourceVersion = ""
+						assert.Equal(t, expectedRole, actualRole)
+					}
+				}
+			} else {
+				t.Errorf("error retrieving roles from client: %v", err)
```
Comment (candita) [inline pkg/operator/controller/configurable-route/controller_test.go:479]: Encapsulate here too.

File: pkg/operator/controller/configurable-route/controller_test.go [hunk 7]
```diff
@@ -0,0 +1,899 @@
...
+							// Value that we want to ignore in the comparison.
+							actualRole.ResourceVersion = ""
+							actualRole.GenerateName = ""
+							expectedRole.GenerateName = ""
+							assert.Equal(t, expectedRole, actualRole)
+						}
+					}
+				} else {
+					t.Errorf("error retrieving roles from client: %v", err)
+				}
```
Comment (candita) [inline pkg/operator/controller/configurable-route/controller_test.go:693]: Encapsulate

**Comment by gcs278** [comment]
> Did you miss adding go.mod and go.sum?

There are no changes to go.mod and go.sum. I didn't add any new Go modules, just used new packages in existing modules. It seems Go only downloads the packages inside the modules that you use, I guess to keep things minimal.

File: pkg/operator/controller/configurable-route/controller_test.go (hunk 1 above)
Comment (gcs278) [reply pkg/operator/controller/configurable-route/controller_test.go:159]: Done.

File: pkg/operator/controller/configurable-route/controller_test.go (hunk 3 above)
Comment (gcs278) [reply pkg/operator/controller/configurable-route/controller_test.go:184]: Good idea. Done.

File: pkg/operator/controller/configurable-route/controller_test.go (hunk 4 above)
Comment (gcs278) [reply pkg/operator/controller/configurable-route/controller_test.go:206]: Done.

File: pkg/operator/controller/configurable-route/controller_test.go (hunk 2 above)
Comment (gcs278) [reply pkg/operator/controller/configurable-route/controller_test.go:214]: We talked a bit in slack about this. `intersectingComponentRoutes` does provide additional value through testing edge cases with component routes. It tests for invalid component routes. I think these are good reasons to unit test and don't think we should remove these test cases.

However, I simply just moved the tests cases to the Reconcile function, since it was just as easy to test them there and removed this unit test. Let me know if you disagree.

File: pkg/operator/controller/configurable-route/controller_test.go (hunk 5 above)
Comment (gcs278) [reply pkg/operator/controller/configurable-route/controller_test.go:380]: I use Goland and it continuously formats the code. Just to be sure, I ran `go fmt` and this is how it likes it.

File: pkg/operator/controller/configurable-route/controller_test.go (hunk 6 above)
Comment (gcs278) [reply pkg/operator/controller/configurable-route/controller_test.go:479]: done

File: pkg/operator/controller/configurable-route/controller_test.go (hunk 7 above)
Comment (gcs278) [reply pkg/operator/controller/configurable-route/controller_test.go:693]: done

File: pkg/operator/controller/configurable-route/controller_test.go [hunk 8]
```diff
@@ -0,0 +1,871 @@
...
+		}
+	} else {
+		t.Errorf("error retrieving roles from client: %v", err)
+	}
+}
+
+// assertRoleBindingsEqual compares provided expectedRoleBindings to the actual roleBindings that exist in the client
+// It ignores generated metadata fields such as ResourceVersion.
+func assertRoleBindingsEqual(t *testing.T, client client.Client, expectedRoleBindings *rbacv1.RoleBindingList) {
+	actualRoleBindings := &rbacv1.RoleBindingList{}
```
Comment (candita) [inline pkg/operator/controller/configurable-route/controller_test.go:848]:
```suggestion
        t.Helper()
        
	actualRoleBindings := &rbacv1.RoleBindingList{}
```

File: vendor/modules.txt [hunk 9]
```diff
@@ -815,6 +815,7 @@ k8s.io/client-go/applyconfigurations/storage/v1beta1
 k8s.io/client-go/discovery
 k8s.io/client-go/discovery/cached
 k8s.io/client-go/discovery/cached/memory
+k8s.io/client-go/discovery/fake
```
Comment (candita) [inline vendor/modules.txt:818]: How does the size of the package change with the additional 225 files?  Does it take longer to download with all the extras?

File: pkg/operator/controller/configurable-route/controller_test.go (hunk 8 above)
Comment (gcs278) [reply pkg/operator/controller/configurable-route/controller_test.go:848]: Good catch, done.

File: vendor/modules.txt (hunk 9 above)
Comment (gcs278) [reply vendor/modules.txt:818]: Looks like it adds about ~1.4 MB to the total size of the repo. We are at 71MB in the vendor files. These days, I'd imagine everyone has fairly high speed networks, so maybe 1 additional second for folks to get this update.

**Comment by gcs278** [comment]
/hold
Some of Miciah's comments in https://github.com/openshift/cluster-ingress-operator/pull/849 also apply here, I'd like to make them consistent.
